### ISSUE TITLE
[Question] TypeScript rewrite

### ISSUE THREAD

**Issue by Lantianyou** [xenova/transformers.js#264]
Hi Joshua. I found your idea is extremely exciting.
I am a frontend developer who has worked on TypeScript professionally for three years. Would you mind me doing a TypeScript re-write, so this npm package can have a better DX.

I just forked this repo. https://github.com/Lantianyou/transformers.js

**Comment by Lantianyou** [comment]
I do not think TS is always the best solution either, however, it would be nice to have it for this project. For example, this project use BigInt, it assumes ES 2020 or later. I think TypeScript transpiler can be a great use case here.

**Comment by quantuminformation** [comment]
TS will die if js ever gets types, its at proposal stage atm

**Comment by kungfooman** [comment]
> I do not think TS is always the best solution either, however, it would be nice to have it for this project.

No, it would be as ugly as it gets, it just makes everything 200 times more complicated while not actually improving anything (actually decreasing the readability) and prevents everyone from quickly testing and making PR's because you have to fumble with a build process in different scenarious, even though you just want to test a few lines of new code.

**Comment by quantuminformation** [comment]
yup, why do you think I dumped ts for my flagship patterns:

https://github.com/quantuminformation/vanillajs-patterns

**Comment by Lantianyou** [comment]
I believe type proposal for ES is in a very far future.

**Comment by sigma-andex** [comment]
I give my 👍 to the _rewrite this in typescript_ proposal. While I agree that TS config can be bit annoying - not to say a pain the a** - I do think that a simple setup is possible that works for most people checking out this repo (at least if you are not using something abstruse like Nix). I think the benefits ( type-safety, better editor auto-completion, using the language that most modern web libraries are written in, ...) far outweigh the config hiccups.

**Comment by Lantianyou** [comment]
Thank you @sigma-andex for your support.

**Comment by kungfooman** [comment]
> I think the benefits ( type-safety, better editor auto-completion, using the language that most modern web libraries are written in, ...) far outweigh the config hiccups.

You are just exposing that you don't know how current technology works. JSDoc and TypeScript is using the *same* language server for IntelliSense.

You gain nothing but have a PITA now. *you are sold*

The amount of TypeScript fanboys that are convinced how much better TS is while not knowing anything about the underlying technology is really getting worrisome...

**Comment by quantuminformation** [comment]
TypeScript is great if you like low signal/noise ratios

**Comment by marek-hanzal** [comment]
Hello - Don't think TS will die even with new JS proposal as TS does much more the proposal solves (and yes, it could be very painfull, but it's basically just about devs skills).

But rewriting the code is not a good idea as it's possible to create and maintain `.d.ts` file, if devs want to use types (for example `@types/tramsformerjs`).

I'm not sure if it makes for this library though...

**Comment by tocque** [comment]
> > I think the benefits ( type-safety, better editor auto-completion, using the language that most modern web libraries are written in, ...) far outweigh the config hiccups.
> 
> You are just exposing that you don't know how current technology works. JSDoc and TypeScript use the _same_ language server for IntelliSense.
> 
> You gain nothing but have a PITA now. _you are sold_
> 
> The amount of TypeScript fanboys that are convinced how much better TS is while not knowing anything about the underlying technology is really getting worrisome...

What you said is absolutely correct in theory.

But I must point out that, based on the actual performance of this repository (and most other repositories that make such claims), when people choose not to use TypeScript, their subtext is that they do not intend to provide particularly good type annotations, and as a result, these libraries have significantly worse IntelliSense experiences.

**Comment by kungfooman** [comment]
> What you said is absolutely correct in theory.
> 
> But I must point out that, based on the actual performance of this repository (and most other repositories that make such claims), when people choose not to use TypeScript, their subtext is that they do not intend to provide particularly good type annotations, and as a result, these libraries have significantly worse IntelliSense experiences.

You like this pattern? [AGREE WITH PROVISO] but [MOTIVATED REASONING]

And I checked some of your projects and your documentation skills are not exactly ground-breaking. You're even using TSDOC incorrectly + non-english comments + no parameter descriptions + no `@example` tags etc. People in glass houses shouldn't throw stones. 😉 

I've seen too many "rewrites" that basically just slap `as any` at everything, because when typing *actually* is required and becomes hard, most TypeScript developers chicken out faster than they would ever admit. This repo *actually* goes several extra-miles to provide *good* typing.

And you know what? If it's not good enough for your refined taste yet, you can make PR's to improve it and you don't even need to break ESM support for that or turning everything into cluttered and unreadable code+type spaghetti! ❣️

**Comment by nico-martin** [comment]
@kungfooman, We are very aware of how important good type support is for our users. That's why we use very comprehensive JSDoc type annotations, which automatically generate d.ts files for all used and exposed functionalities and are included with the library.
A complete rewrite to TypeScript is not planned at this time, as we do not believe the benefits justify the effort.

However, if you encounter type issues with specific functions, we would be happy to look into them.

Btw, v4 brings a huge number of type improvements.

**Comment by kungfooman** [comment]
> [@kungfooman](https://github.com/kungfooman), We are very aware of how important good type support is for our users. That's why we use very comprehensive JSDoc type annotations, which automatically generate d.ts files for all used and exposed functionalities and are included with the library.

Yea, I also love good typing, that's why I contributed the initial type declaration generation to this repo three years ago 😉

**Comment by nico-martin** [comment]
Sorry, I guess I should have tagged @tocque in my reply 😊.
@kungfooman , thank you so much for your contributions!
